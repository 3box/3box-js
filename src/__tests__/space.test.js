jest.mock('../keyValueStore')
const ENSURE_CONNECTED = 'ensure connected function'
const ORBITDB = 'orbitdb instance'
const threeIdMock = {
  initKeyringByName: jest.fn(() => {
    return false
  }),
  getKeyringBySpaceName: jest.fn(() => {
    return {
      getDBSalt: jest.fn(),
      symEncrypt: data => { return { ciphertext: 'wow, such encrypted/' + data, nonce: 123 } },
      symDecrypt: data => data.split('/')[1]
    }
  })
}
let rootstoreMockData = []
const rootstoreMock = {
  iterator: () => { return { collect: () => rootstoreMockData } },
  add: jest.fn()
}

const Space = require('../space')


describe('Space', () => {

  let space
  let NAME1 = 'test1'
  let NAME2 = 'test2'

  beforeEach(() => {
    rootstoreMock.add.mockClear()
    threeIdMock.initKeyringByName.mockClear()
    threeIdMock.getKeyringBySpaceName.mockClear()
  })

  it('should be correctly constructed', async () => {
    space = new Space(NAME1, threeIdMock, ORBITDB, rootstoreMock, ENSURE_CONNECTED)
    expect(space._name).toEqual(NAME1)
    expect(space._3id).toEqual(threeIdMock)
    expect(space._rootStore).toEqual(rootstoreMock)
    expect(space._store._orbitdb).toEqual(ORBITDB)
    expect(space._store._ensureConnected).toEqual(ENSURE_CONNECTED)
    expect(space._store._name).toEqual('3box.space.' + NAME1 + '.keyvalue')
  })

  it('should open a new space correctly', async () => {
    let opts = {
      consentCallback: jest.fn(),
    }
    const syncDonePromise = new Promise((resolve, reject) => {
      opts.onSyncDone = resolve
    })
    await space.open(opts)
    expect(opts.consentCallback).toHaveBeenCalledWith(false, NAME1)
    expect(rootstoreMock.add).toHaveBeenCalledWith({ odbAddress:'/orbitdb/myodbaddr' })
    expect(threeIdMock.initKeyringByName).toHaveBeenCalledWith(NAME1)
    expect(threeIdMock.getKeyringBySpaceName).toHaveBeenCalledWith('3box.space.' + NAME1 + '.keyvalue')
    await syncDonePromise
  })

  it('should return directly if space already opened', async () => {
    let opts = {
      consentCallback: jest.fn(),
    }
    await space.open(opts)
    expect(opts.consentCallback).toHaveBeenCalledTimes(0)
  })

  it('should open correctly and not add to rootStore if entry already present', async () => {
    rootstoreMockData.push({ payload: { value: { odbAddress: '/orbitdb/Qmasofgh/3box.space.' + NAME2 + '.keyvalue'} } })
    let opts = {
      consentCallback: jest.fn(),
    }
    const syncDonePromise = new Promise((resolve, reject) => {
      opts.onSyncDone = resolve
    })
    space = new Space(NAME2, threeIdMock, ORBITDB, rootstoreMock, ENSURE_CONNECTED)
    await space.open(opts)
    expect(opts.consentCallback).toHaveBeenCalledWith(false, NAME2)
    expect(rootstoreMock.add).toHaveBeenCalledTimes(0)
    expect(threeIdMock.initKeyringByName).toHaveBeenCalledWith(NAME2)
    expect(threeIdMock.getKeyringBySpaceName).toHaveBeenCalledWith('3box.space.' + NAME2 + '.keyvalue')
    await syncDonePromise
  })

  describe('public store reducer', () => {
    it('get/set/remove works correctly', async () => {
      await space.public.set('k1', 'v1')
      await space.public.set('k2', 'v2')
      await space.public.set('k3', 'v3')
      expect(await space.public.get('k1')).toEqual('v1')
      expect(await space.public.get('k2')).toEqual('v2')
      expect(await space.public.get('k3')).toEqual('v3')
      await space.public.remove('k2')
      expect(await space.public.get('k2')).toBeUndefined()
    })

    it('log should only return public values', async () => {
      const refLog = [ { key: 'k1', op: 'PUT', timeStamp: 123, value: 'v1' }, { key: 'k3', op: 'PUT', timeStamp: 123, value: 'v3' } ]
      const log1 = space.public.log
      expect(log1).toEqual(refLog)
      space._store.set('key', 'value')
      const log2 = space.public.log
      expect(log2).toEqual(log1)
    })

    it('all should return all values from public store', async () => {
      expect(await space.public.all()).toEqual({
        k1: 'v1',
        k3: 'v3'
      })
    })
  })

  describe('private store reducer', () => {
    it('get/set/remove works correctly', async () => {
      await space.private.set('k1', 'sv1')
      await space.private.set('k2', 'sv2')
      await space.private.set('k3', 'sv3')
      expect(await space.private.get('k1')).toEqual('sv1')
      expect(await space.private.get('k2')).toEqual('sv2')
      expect(await space.private.get('k3')).toEqual('sv3')
      await space.private.remove('k2')
      expect(await space.private.get('k2')).toEqual(null)
    })

    it('log should only return private values', async () => {
      const refLog = [ { key: 'k1', op: 'PUT', timeStamp: 123, value: 'sv1' }, { key: 'k3', op: 'PUT', timeStamp: 123, value: 'sv3' } ]
      const log1 = space.private.log
      expect(log1).toEqual(refLog)
      space._store.set('key', 'value')
      const log2 = space.private.log
      expect(log2).toEqual(log1)
    })

    it('all should return all values from private store', async () => {
      expect(await space.private.all()).toEqual({
        k1: 'sv1',
        k3: 'sv3'
      })
    })
  })
})
