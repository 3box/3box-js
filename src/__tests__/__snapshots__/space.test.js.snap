// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Space Threads does not subscribe or return invalid thread address (ignore experimental) 1`] = `"subscribeThread: must subscribe to valid thread/orbitdb address"`;

exports[`Space public store reducer all should return all values from public store 1`] = `
Object {
  "k1": "v1",
  "k3": "v3",
  "k4": "v4",
  "k5": "v5",
}
`;

exports[`Space public store reducer all should return all values from public store 2`] = `
Object {
  "k1": Object {
    "timestamp": 1558954408,
    "value": "v1",
  },
  "k3": Object {
    "timestamp": 1558954408,
    "value": "v3",
  },
  "k4": Object {
    "timestamp": 1558954408,
    "value": "v4",
  },
  "k5": Object {
    "timestamp": 1558954408,
    "value": "v5",
  },
}
`;

exports[`Space public store reducer log should only return public values 1`] = `
Array [
  Object {
    "key": "k1",
    "op": "PUT",
    "timeStamp": 123,
    "value": "v1",
  },
  Object {
    "key": "k3",
    "op": "PUT",
    "timeStamp": 123,
    "value": "v3",
  },
  Object {
    "key": "k4",
    "op": "PUT",
    "timeStamp": 123,
    "value": "v4",
  },
  Object {
    "key": "k5",
    "op": "PUT",
    "timeStamp": 123,
    "value": "v5",
  },
]
`;
